File: 1-C/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long cardnumber = get_long("%s", "Number: ");
    long workingnumber = cardnumber;
    int digits = 0; // Will hold the total number of digits
    int lastdigit; // Will hold the most recent digit
    int secondlastdigit; // Will hold the digit before that
    int checksum = 0; // Working sum

    while (workingnumber) // Will end when workingnumber hits zero
    {
        digits += 1;
        int thisdigit = workingnumber % 10; // Get the smallest digit
        secondlastdigit = lastdigit; // Push stored digit back one
        lastdigit = thisdigit; // Store new digit

        if (digits % 2 == 1) // If digit is odd
        {
            checksum += thisdigit; // Straight add to the checksum
        }
        else // If digit is even
        {
            checksum += ((thisdigit * 2) % 10) + ((thisdigit * 2) / 10); // Add the individual digits of thisdigit * 2
        }

        workingnumber /= 10; // Chop a digit off the end of workingnumber
    }

    if (checksum % 10 == 0)
    {
        if ((digits == 13 || digits == 16) && lastdigit == 4)
        {
            printf("%s", "VISA\n");
        }

        else if (digits == 15  && lastdigit == 3 && (secondlastdigit == 4 || secondlastdigit == 7))
        {
            printf("%s", "AMEX\n");
        }
        else if (digits == 16  && lastdigit == 5 && secondlastdigit >= 1 && secondlastdigit <= 5)
        {
            printf("%s", "MASTERCARD\n");
        }
        else
        {
            printf("%s", "INVALID\n");
        }

    }
    else
    {
        printf("%s", "INVALID\n");
    }

}

File: 6-Python/sentimental-credit/credit.py
def main():
    cardnumber = input("Number: ")
    if check_sum(cardnumber):
        print(f"{check_number(cardnumber)}\n")
    else:
        print("INVALID\n")


def check_sum(cardnumber):
    checksum = 0
    even = True
    for i in range(len(cardnumber), 0, -1):
        try:
            thisdigit = int(cardnumber[i - 1])
        except:
            return False

        if even:  # Add the current digit
            checksum += thisdigit
        else:  # Add the individual digits of the current digit * 2
            checksum += int((thisdigit * 2) % 10)
            checksum += int((thisdigit * 2) / 10)
        even = not even

    if checksum % 10 == 0:
        return True
    else:
        return False


def check_number(cardnumber):
    number_length = len(cardnumber)
    if number_length in (13, 16) and int(cardnumber[0]) == 4:
        return "VISA"
    elif number_length == 15 and int(cardnumber[0:2]) in (34, 37):
        return "AMEX"
    elif number_length == 16 and 51 <= int(cardnumber[0:2]) <= 55:
        return "MASTERCARD"
    else:
        return "INVALID"


if __name__ == "__main__":
    main()
